server/punishments.ts

```diff
@@ -861,17 +864,24 @@ export const Punishments = new class {
 		}
 		return success;
 	}
-	automaticUnlock(user: User, punishedID: ID | PunishType) {
-		if (user.locked === punishedID) {
-			Punishments.unlock(user.id);
+	/**
+	 * Checks if a punishment has expired, and unlocks it if so. Otherwise, pokes the punishment timer.
+	 */
+	runExpiry(user: User, punishment: Punishment) {
+		const [, id, expireTime] = punishment;
+		if (expireTime < Date.now()) {
+			if (user.locked === id) Punishments.unlock(user.id);
+		} else {
+			Punishments.pokePunishmentTimer(user, punishment);
 		}
 	}
 	/**
-	 * Validates the length and sets the punishment timer for a user.
+	 * Sets the punishment timer for a user,
+	 * to either MAX_PUNISHMENT_TIMER_LENGTH or the amount of time left on the punishment.
 	 */
-	setPunishmentTimer(user: User, expireTime: number, punishedID: ID | PunishType) {
-		if (expireTime === 0 || expireTime === Infinity || expireTime > Chat.MAX_TIMEOUT_DURATION) return;
-		user.punishmentTimer = setTimeout(Punishments.automaticUnlock, expireTime - Date.now(), user, punishedID);
+	pokePunishmentTimer(user: User, punishment: Punishment) {
+		const length = Math.min(punishment[2] - Date.now(), MAX_PUNISHMENT_TIMER_LENGTH);
+		user.punishmentTimer = setTimeout(Punishments.runExpiry, length, user, punishment);
```

Comment:
I never like `setTimeout` with arguments, I think it has worse readability.

```suggestion
		user.punishmentTimer = setTimeout(() => {
			Punishments.runExpiry(user, punishment);
		}, length);
```

On the other hand, your approach _is_ what let the previous one hotpatch correctly. Still, though. Probably better just to `clearTimeout`.

Also, why isn't `pokePunishmentTimer` and `runExpiry` the same function? I'd call the merged function `checkPunishmentTime`.